crates/server/src/server/server_future.rs

```diff
@@ -642,71 +672,141 @@ impl<T: RequestHandler> ServerFuture<T> {
-                    if let Err(e) = sanitize_src_address(src_addr) {
-                        warn!(
-                            "address can not be responded to {src_addr}: {e}",
-                            src_addr = src_addr,
-                            e = e
-                        );
-                        continue;
-                    }
-
-                    let handler = handler.clone();
-                    let dns_hostname = dns_hostname.clone();
+                    },
+                    _ = drain.clone() => {
+                        // A graceful shutdown was initiated. Break out of the loop.
+                        break;
+                    },
+                };
+
+                // verify that the src address is safe for responses
+                // TODO: we're relying the quinn library to actually validate responses before we get here, but this check is still worth doing
+                if let Err(e) = sanitize_src_address(src_addr) {
+                    warn!(
+                        "address can not be responded to {src_addr}: {e}",
+                        src_addr = src_addr,
+                        e = e
+                    );
+                    continue;
+                }
 
-                    inner_join_set.spawn(async move {
-                        debug!("starting quic stream request from: {src_addr}");
+                let handler = handler.clone();
+                let dns_hostname = dns_hostname.clone();
 
-                        // TODO: need to consider timeout of total connect...
-                        let result = quic_handler(handler, streams, src_addr, dns_hostname).await;
+                inner_join_set.spawn(async move {
+                    debug!("starting quic stream request from: {src_addr}");
 
-                        if let Err(e) = result {
-                            warn!("quic stream processing failed from {src_addr}: {e}")
-                        }
-                    });
+                    // TODO: need to consider timeout of total connect...
+                    let result =
+                        quic_handler(handler, streams, src_addr, dns_hostname, drain.clone()).await;
 
-                    reap_tasks(&mut inner_join_set);
-                }
+                    if let Err(e) = result {
+                        warn!("quic stream processing failed from {src_addr}: {e}")
+                    }
+                });
             }
+
+            reap_tasks(&mut inner_join_set);
+            Ok(())
         });
 
         Ok(())
     }
 
-    /// This will run until a background task of the trust_dns_server ends.
+    /// Triggers a graceful shutdown the server. All background tasks will stop accepting
+    /// new connections and the returned future will complete once all tasks have terminated.
+    pub fn shutdown(self) -> impl Future<Output = Result<(), ProtoError>> {
```

Comment:
Alternatively, we could return both the future and a drain handle, to allow the application to handle them separately. I'm open to suggestions.